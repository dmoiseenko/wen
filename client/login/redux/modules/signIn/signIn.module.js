import { createAction } from 'redux-act';


export const signIn = createAction(
  'sign in'
);

export const signInSuccess = createAction(
  'sign in success'
);

export const signInFailure = createAction(
  'sign in failure'
);

export const reducer = {
  [signIn]: state => ({
    ...state,
    signIn: {
      ...state.signIn,
      loading: true,
      loaded: false
    }
  }),

  [signInSuccess]: state => ({
    ...state,
    signIn: {
      ...state.signIn,
      loading: false,
      loaded: true
    }
  }),

  [signInFailure]: state => ({
    ...state,
    signIn: {
      ...state.signIn,
      loading: false,
      loaded: false
    }
  })
};
