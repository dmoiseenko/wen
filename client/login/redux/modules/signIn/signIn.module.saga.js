import { call, put, select, takeEvery } from 'redux-saga/effects';
import { startSubmit, stopSubmit } from 'redux-form';

import * as api from '../../../core/api';
import * as signInSelector from '../../../core/signIn.selector';
import * as signInForm from '../../../core/form.constants';
import * as location from '../../../../common/core/location';
import * as routes from '../../../../../common/routes';
import * as module from './signIn.module';


export function* signInSuccessSaga(response) {
  yield put.resolve(module.signInSuccess(response));
  yield put(stopSubmit(signInForm.signInForm));
  yield call(location.setLocation, routes.home);
}

export function* signInFailureSaga(error) {
  yield put.resolve(module.signInFailure(error.error));
  yield put(stopSubmit(signInForm.signInForm, {
    [signInForm.emailField]: error.error,
    [signInForm.passwordField]: error.error
  }));
}

export function* getEmailAndPasswordSaga() {
  const email = yield select(signInSelector.email);
  const password = yield select(signInSelector.password);

  return { email, password };
}

export function* signInSaga() {
  const { email, password } = yield call(getEmailAndPasswordSaga);

  yield put(startSubmit(signInForm.signInForm));

  const { response, error } = yield call(api.login, { email, password });

  if (response) {
    yield call(signInSuccessSaga, response);
  } else {
    yield call(signInFailureSaga, error);
  }
}

export function* watchSignInSaga() {
  yield takeEvery(module.signIn.getType(), signInSaga);
}
