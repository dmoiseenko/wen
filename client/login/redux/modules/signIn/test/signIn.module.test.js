import deepFreeze from 'deep-freeze';

import * as module from '../signIn.module';


describe('reducer', () => {
  const stateBefore = {
    signIn: {}
  };

  deepFreeze(stateBefore);

  it('signIn', () => {
    const actual = module.reducer[module.signIn](stateBefore);

    expect(actual).toMatchSnapshot();
  });

  it('signInSuccess', () => {
    deepFreeze(stateBefore);

    const actual = module.reducer[module.signInSuccess](stateBefore);

    expect(actual).toMatchSnapshot();
  });

  it('signInFailure', () => {
    const actual = module.reducer[module.signInFailure](stateBefore);

    expect(actual).toMatchSnapshot();
  });
});
