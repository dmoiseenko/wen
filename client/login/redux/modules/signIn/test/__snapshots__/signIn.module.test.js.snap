// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reducer signIn 1`] = `
Object {
  "signIn": Object {
    "loaded": false,
    "loading": true,
  },
}
`;

exports[`reducer signInFailure 1`] = `
Object {
  "signIn": Object {
    "loaded": false,
    "loading": false,
  },
}
`;

exports[`reducer signInSuccess 1`] = `
Object {
  "signIn": Object {
    "loaded": true,
    "loading": false,
  },
}
`;
